### DISCUSSION TITLE
consult--source-tab

### DISCUSSION THREAD

**Discussion by ezemtsov** [minad/consult#1135] [Q&A]
Hi @minad,
Thanks a lot for your tremendous effort on consult and keeping exwm alive!

At some point I switched to consult-buffer as a single control interface for tabs (which I use instead of workspaces), buffers and exwm windows. It also replaced my linux application launcher:
```elisp
  (consult-buffer-sources
   '((:name "Tabs"
      :category 'tab
      :items (lambda () (mapcar #'(lambda (tab) (alist-get 'name tab)) (tab-bar-tabs)))
      :action (lambda (cand) (tab-switch cand)))
     consult--source-hidden-buffer
     consult--source-modified-buffer
     consult--source-buffer
     consult--source-recent-file
     (:name "Apps"
      :category app
      :items executable-list
      :action (lambda (cand) (start-process cand nil cand)))))
```

While this works really good already, I was wondering how hard would it be to implement preview functionality for the Tabs source?

**Comment by minad** [comment]
> While this works really good already, I was wondering how hard would it be to implement preview functionality for the Tabs source?

Certainly not easy, and I think it is also unnecessary, at least in my workflow. With EXWM I use `tab-bar-mode` which shows me a list of tabs (dedicated separate desktops). The tab bar also shows my global system statusbar. In this setting, preview would not help me much - it would be rather confusing, since I try to isolate buffers for each tab. See also https://github.com/minad/consult/pull/1009. Maybe these ideas are helpful for you?

I wonder which work flow do you use? Why do you want tab preview in the context of EXWM?

**Reply by ezemtsov** [reply]
Thinking more about it over night, I think the only reason that's important for me is because I want to use `consult-buffer` as an application switcher for buffers already opened in another tabs, instead of displaying the selected buffer again in the current tab. 

This is already implemented in Rofi (combi mode), which I used with i3wm before. The idea there is that you have a single application working as a window switcher and as an application launcher, with priority to switching.

Reading the #1009, I think the goal is different there, as it restricts the list of choices only to current tab. This has downsides that a user **must** remember which tab has a certain buffer, which is usually not the case when applications are not bound to specific tab numbers. Instead, my proposed workflow considers having all buffers (hidden and opened) available from anywhere with default actions for:
* **Buffer is hidden** - swap current buffer to the selected hidden buffer (current default behavior)
* **Buffer is opened in some tab** - switch to relevant tab, focus to selected buffer

Just to summarize on the workflow again: if I have a buffer opened in another tab, I want to switch to that tab and focus on selected buffer. That, instead of default behavior of opening the buffer again in current tab in the current buffer. 

Partially I was able to implement this workflow by just having tabs switching prioritized over buffers. By default tab names are just a collection of it's opened buffer names, which as a side-effect makes the workflow possible. So for example:
1. There are N tabs opened, one of them (e.g. number 5) has Telegram opened
2. Currently I'm working in the first tab, where I have some project code and compilation logs
3. While staying in the first tab, I use `consult-buffer` to switch to between various project buffers
4. Dunst shows a notification, pointing that I have a new message from a friend in Telegram
5. I use `consult-buffer` and type `tele` and press enter to switch to the tab with Telegram opened
6. After I replied to my friend I press `mod 1` to switch to the first tab, where I have my project code

**Reply by minad** [reply]
Thanks. I see. Right now I use a two step switching, first select the tab and then switch. This way I get faster local buffer switching in the current tab. But my workflow is certainly not yet optimal and from time to time, I accidentally move buffers between tabs. 

I am not sure if Consult should offer tab switching functionality yet, or something like #1009. The problem is that there are too many alternative workflows and buffer isolation techniques. However if we figure out sufficiently general building blocks usable in multiple scenarios, that would be worth a consideration.

**Reply by minad** [reply]
@ezemtsov I've added some support for tabs. Preview will happen in the current tab, but the tab buffer list is restored afterwards: https://github.com/minad/consult/commit/0a5af92d60efc8bb5727b692f87a0ec042bbf5e8

**Reply by ezemtsov** [reply]
@minad I ended up writing a custom source for tabs that supports preview. In case if it's useful for anyone:

```elisp
;;; -*- lexical-binding: t -*-
(require 'consult)

(consult--define-state tab)

(defun consult--tab-preview ()
  "Tab preview function"
  (let ((orig-tab (alist-get 'name (tab-bar--current-tab))))
    (lambda (action cand)
      (pcase action
        ('return
         (tab-bar-switch-to-tab orig-tab))
        ('exit
         (tab-bar-switch-to-tab orig-tab))
        ('preview
         (if cand (tab-bar-switch-to-tab cand)
           (tab-bar-switch-to-tab orig-tab)))))))

(defun consult--tab-action (cand)
  (tab-bar-switch-to-tab cand))

(defun consult--tab-items ()
  "List of tabs available in the current frame.
   Note, that current tab is always listed first, so that live
   preview wouldn't randomly switch to the first index."
  (let ((tabs (mapcar #'(lambda (tab) (alist-get 'name tab)) (tab-bar-tabs)))
        (current (alist-get 'name (tab-bar--current-tab))))
    (cons current (remove current tabs))))

(defvar consult--source-tab
  `( :name          "Tabs"
     :narrow        ?t
     :category      tab
     :require-match t
     :state         ,#'consult--tab-state
     :enabled       ,(lambda () (if tab-bar-mode t nil))
     :items         ,#'consult--tab-items
     "Tab candidate source for `consult-buffer'."))

(provide 'consult-ext)
```

**Reply by minad** [reply]
Can you please put this in the wiki? Iirc there is already a tab source but it helps to offer multiple variants.

**Reply by ezemtsov** [reply]
After using it for few days, I'm not confident to recommend this to anyone yet.
The issue is that sometimes it changes the screen content so quickly and dramatically that it's hard to follow what exactly happened.
So I end up spending few seconds on checking the state of the tab bar just to understand where I ended up.

I'll spend a bit more time tinkering until it works the way I want.
